elastic/cloud-on-k8s
# Allow users to provide SANs to inject in the node certs

This commit extends the Elasticsearch CRD with the following structure:

```
  tls:
    subjectAltNames:
      - ip: 4.3.2.1
      - ip: 4.3.2.2
      - dns: my.domain.name.com
      - dns: my.second.domain
```

On reconciliation, the operator makes sure these user-provided SANs are
injected into all nodes certificates.
Since ES supports hot-reloading certificates, it's fairly quick to
reload SANs by reapplying a different ES spec.

IP addresses are validated in our ES spec validation handler, ipv6 is
supported.

This fixes cert verification on public deployments: before this commit
users would have a hard time exposing their ES cluster on any IP/DNS
outside the k8s cluster.

A way to manually test this:
* deploy a sample ES cluster, exposed through a load balancer
* grab the CA cert secret
* try to curl ES through the load balancer public endpoint with the CA
cert: it's impossible to verify the server cert since hostname mismatch
* add the load balancer public IP in the ES spec
* perform the same curl commands: cert is correctly validated

Follow-up PR to setup an E2E test for it: https://github.com/elastic/k8s-operators/issues/724

Fixes #713.
